composition-js/src/merging/merge.ts

```diff
@@ -1102,40 +1106,55 @@ class Merger {
-    if (useIntersection) {
-      for (const source of sources) {
-        if (!source) {
+      source.arguments().forEach((arg) => argNames.add(arg.name));
+    }
+
+    for (const argName of argNames) {
+      // We add the arguement unconditionally even if we're going to remove it in
+      // some path. Done because this helps reusing our "reportMismatchHint" method
+      // in those cases.
+      const arg = dest.addArgument(argName);
+      // If all the sources that have the field have the argument, we do merge it
+      // and we're good, but otherwise ...
+      if (!sources.every((s) => !s || s.argument(argName))) {
+        // ... we don't merge the argument: some subgraphs wouldn't know what
+        // to make of it and that would be dodgy at best. If the argument is
+        // optional in all sources, then we can compose properly and just issue a
+        // hint. But if it is mandatory, then we have to fail composition, otherwise
+        // the query planner would have no choice but to generate invalidate queries.
+        const nonOptionalSources = sources.map((s, i) => s && s.argument(argName)?.isRequired() ? this.names[i] : undefined).filter((s) => !!s) as string[];
+        if (nonOptionalSources.length > 0) {
+          const nonOptionalSubgraphs = printSubgraphNames(nonOptionalSources);
+          const missingSources = printSubgraphNames(sources.map((s, i) => s && !s.argument(argName) ? this.names[i] : undefined).filter((s) => !!s) as string[]);
+          this.errors.push(ERRORS.REQUIRED_ARGUMENT_MISSING_IN_SOME_SUBGRAPH.err({
+            message: `Argument "${arg.coordinate}" is required in some subgraphs but does not appear in all subgraphs: it is required in ${nonOptionalSubgraphs} but does not appear in ${missingSources}`,
+            nodes: sourceASTs(...sources.map((s) => s?.argument(argName))),
+          }));
+        } else {
           this.reportMismatchHint(
             hintInconsistentArgumentPresence,
-            `Argument "${dest.coordinate}" will not be added to "${dest.parent}" in the supergraph as it does not appear in all subgraphs: `,
-            dest,
-            sources,
+            `Argument "${arg.coordinate}" will not be added to "${dest}" in the supergraph as it does not appear in all subgraphs: `,
+            arg,
+            sources.map((s) => s ? s.argument(argName) : undefined),
             _ => 'yes',
             // Note that the first callback is for element that are "like the supergraph" and we've pass `dest`.
             (_, subgraphs) => `it is defined in ${subgraphs}`,
             (_, subgraphs) => ` but not in ${subgraphs}`,
             undefined,
             true // Do include undefined sources, that's the point
           );
-          // Note that we remove the element after the hint because we access the parent in the hint message.
-          dest.remove();
-          return;
         }
+        arg.remove();
```

Comment:
There was a comment about why this was being removed that helped understanding of the code. Could you add it back in?

Reply:
Added back.